ring/lifecycler.go

```diff
@@ -593,11 +593,46 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
 			return ringDesc, true, nil
 		}
 
-		// If the ingester failed to clean its ring entry up in can leave its state in LEAVING
+		// If the ingester failed to clean its ring entry up it can leave its state in LEAVING
 		// OR unregister_on_shutdown=false
 		// Move it into ACTIVE to ensure the ingester joins the ring.
-		if instanceDesc.State == LEAVING && len(instanceDesc.Tokens) == i.cfg.NumTokens {
-			instanceDesc.State = ACTIVE
+		if instanceDesc.State == LEAVING {
+			var tokens Tokens = instanceDesc.Tokens // way of forcing tokens to be of type Tokens instead of []uint32.
+			setIsActive := true
+			if len(instanceDesc.Tokens) != i.cfg.NumTokens {
+				level.Debug(i.logger).Log("msg", "existing entry has different number of tokens", "existingTokens", len(instanceDesc.Tokens), "newTokens", i.cfg.NumTokens)
+				if len(tokensFromFile) > 0 {
+					level.Debug(i.logger).Log("msg", "adding tokens from file", "tokens", len(tokensFromFile))
+					setIsActive = len(tokensFromFile) >= i.cfg.NumTokens
```

Comment:
Why does the instance become active only when the number of tokens in the file are at least i.cfg.NumTokens?
What if `len(tokensFromFile) < i.cfg.NumTokens`, shouldn't the needed tokens be generated and added to the existing ones like we do later, and then set it to `ACTIVE`? The same when `len(tokensFromFile) > i.cfg.NumTokens`, shouldn't the extra tokens be removed?

Reply:
Thanks @agardiman, please see revised/simplified implementation :)